### ISSUE TITLE
retry of completely new chunked upload

### ISSUE THREAD

**Issue by moscicki** [owncloud/client#2667]
Hello,

We think that currently there is a design flaw in 1.7 such that when one chunk PUT fails due to the server entering into a funny state (e.g. out of storage and forgets other already uploaded chunks) the client would be stuck forever (retrying just the failed chunk).

However, in the master branch the logic has been fixed and the client would fail the whole chunked transfer and start over with uploading all chunks. Lines 449-451 in
https://github.com/owncloud/client/blob/master/src/libsync/propagateupload.cpp

Could you please confirm 
1) if our analysis is correct
2) if the new chunked PUT logic is an intended new feature in 1.8

Thanks,

kuba

**Comment by moscicki** [comment]
3) Would the transfer id be different when the new chunked upload on the same file is tried?

**Comment by guruz** [comment]
@moscicki btw, you can link to a line number by clicking it on the left

@ogoffart (this is simnilar to what we have talked about before christmas)

**Comment by ogoffart** [comment]
Normaly, 1.7 and master are not different in that respect:  They should retry to send from the sent chunk, then start over sending the first chunks if the server do not aknoweldge the transfer.

Master contains code that allow to optionally transfer the chunk for the same file in parallel. We do not know yet if we should enable that or not.
But that should not affect the behaviour in case of error.

**Comment by moscicki** [comment]
Thx, here is the line: https://github.com/owncloud/client/blob/master/src/libsync/propagateupload.cpp#L449

**Comment by moscicki** [comment]
@ogoffart: what exactly is the behviour? Would the client try 3 times to PUT the chunk and then give up and start over the whole new transfer with the new transfer id?

**Comment by ogoffart** [comment]
it keeps the same transfer id:
Imagine the chunk 5 (out of 10) fails.
Next sync, it will retry from chunk 5 and send the chunk from 5 to 10.  But then after having sent the chunk 10, the server did not accept the file (by returning an etag).  So the client will continue sending chunks 0, 1, 2, 3 and 4  (with the same transfer id).   Normaly, after the chunk 4, the server should accept the transfer. If it doesn't, the client throw an error.

**Comment by moscicki** [comment]
Okey, then in the last case (client throws an error) the user may manually press the "Resync" button, right? In that case the new chunked transfer with the new transfer id would be started?

**Comment by moscicki** [comment]
@ogoffart: could you please confirm that by pressing "resync" button the client will try the new chunked upload with the new transfer id (see previous question).

**Comment by ogoffart** [comment]
no, a new transfer id is not started with the "resync" button.
The resync button only clears the blacklist.

**Comment by ogoffart** [comment]
@moscicky Do you have more question or can this issue be closed?

**Comment by moscicki** [comment]
@ogoffart: thanks for the answers. I think this protocol may/should be improved so that the client may start a brand new upload in case the server enters an unrecoverable state with the current transfer id. Currently there is no way of doing that and we have seen some problems with that. I guess it could be added in a simple way with some error response code (from top of my head, for example, 424 Failed Dependency).  

I would like to propose this: how should I proceed?

**Comment by ogoffart** [comment]
I fail to understand the use case.
The server can just forget anything about that transfer id.

The next sync will not send the chunk in order but that should not be a problem.

**Comment by moscicki** [comment]
The problem is that this scheme of client/server interaction is not a proper protocol but it is (ab)using knowledge of implicit behaviour of chunk upload implementation on the owncloud server (which btw is extremely inefficient as it needs to write and read the file content twice by writing and concatenating chunks). You can easily think of much more robust implementation of the server which stores the chunks directly into the file offsets. That's not possible however if there is no proper protocol which would at least provide some error recovery signaling between the client and the server. So you are just stuck to one, particular and very inefficient implementation.

The use case is that opencloudmesh initiative (started by Owncloud Inc.) tries to promote de-facto standards for sync&share, such as the webdav based sync protocol, which many people would like to see become a reality. I think this is something that you guys should discuss. Ultimately coding against a well specified protocol should be an advantage also the developers.

**Comment by ogoffart** [comment]
I don't see how it is a problem.  The chunks can come in any order, and even in parallel.
This is why the client also put the chunk size as an header, so the server can write the chunk in place because it knows that all chunk but the last one will have the same size.

**Comment by ogoffart** [comment]
I agree coding against an open protocol is good. But as long as there is no open protocol to follow, this is out of scope of this bug tracker.

I understand you are trying to make such open protocol and i'd be glad to help.

**Comment by moscicki** [comment]
That's really great! We have started writing it up, of course this is very preliminary: https://github.com/cernbox/smashbox/tree/master/protocol

It does not cover the error handling cases yet (like the one discussed in this issue).
